runtime/Java/src/org/antlr/v4/runtime/atn/ATNSerializer.java

```diff
@@ -98,20 +97,15 @@ public IntegerList serialize() {
 				precedenceStates.add(s.stateNumber);
 			}
 
-			data.add(stateType);
+			writer.writeCompactUInt32(stateType);
 
-			if (s.ruleIndex == -1) {
-				data.add(Character.MAX_VALUE);
-			}
-			else {
-				data.add(s.ruleIndex);
-			}
+			writer.writeUInt16(s.ruleIndex == -1 ? Character.MAX_VALUE : s.ruleIndex);
```

Comment:
Unfortunately, it doesn't allow more than 65534 tokens. It looks like some applications require more than 65534 tokens, see https://github.com/antlr/antlr4/issues/840#issuecomment-583880796:

> Sorry to raise this question again! I am trying to build a large dictionary of about 200,000 words (which use simple regular expressions to cover for edge cases. Unfortunately, I am running into the mentioned issue as well. From what I understand, this is due to the reason that the number of internal ATN states is limited to MAX_VALUE = '\uFFFF'. Is there a recommended way of using the ANTLR lexer as a tagger for large dictionaries.

But with the [new encoding format](https://github.com/antlr/antlr4/discussions/3494) it will be possible.

Reply:
Ok, it looks like there is a way to support such cases with saving backward compatibility. I'll try to implement it (but anyway it's quite dirty).